actions/send-ci-cd-trace/src/index.ts

```diff
@@ -0,0 +1,154 @@
+      .setup(connectionString)
+      .setAutoCollectRequests(true)
+      .setAutoCollectPerformance(true, false)
+      .setAutoCollectExceptions(true)
+      .setAutoCollectDependencies(true)
+      .setAutoCollectConsole(true, false)
+      .setAutoCollectPreAggregatedMetrics(true)
+      .setSendLiveMetrics(false)
+      .setInternalLogging(false, true)
+      .enableWebInstrumentation(false)
+      .start();
+
+    const client = appInsights.defaultClient;
+    client.config.aadTokenCredential = credential;
+
+    // Get GitHub context
+    const { repo, runId, job } = github.context;
+    const octokit = github.getOctokit(token);
+
+    // Fetch workflow run details from GitHub API
+    const getWorkflowRunResponse = await octokit.rest.actions.getWorkflowRun({
+      owner: repo.owner,
+      repo: repo.repo,
+      run_id: runId,
+    });
+
+    const workflowRun = getWorkflowRunResponse.data;
+
+    // Fetch jobs related to this workflow run
+    const jobsResponse = await octokit.rest.actions.listJobsForWorkflowRun({
+      owner: repo.owner,
+      repo: repo.repo,
+      run_id: runId,
+    });
+
+    const jobs = jobsResponse.data.jobs;
+    
+    // Check if the current job is the last job in the workflow
+    const isLastJob = jobs[jobs.length - 1]?.name === job;
+
+    // Collect workflow metadata
+    const workflowMetadata = {
+      app: repo.repo,
+      repository: repo.repo,
+      run_id: runId,
+      workflow_name: workflowRun.name,
+      status: workflowRun.status,
+      run_started_at: workflowRun.run_started_at,
+      created_at: workflowRun.created_at,
+      updated_at: workflowRun.updated_at,
+    };
+
+    // Track the workflow start trace
+    const workflowStartTime = new Date(workflowRun.run_started_at || workflowRun.created_at || 0);
+    client.trackEvent({
+      name: `Workflow: ${workflowRun.name}`,
+      properties: {
+        ...workflowMetadata,
+        startTime: workflowStartTime.toISOString(),
+      },
+    });
+
+    // Iterate over the jobs and track each job's information
+    for (const job of jobs) {
+      // Skip the job if it is not the last one
+      if (!isLastJob) {
+        continue;
+      }
+      const jobMetadata = {
+        job_name: job.name,
+        status: job.conclusion,
+        started_at: job.started_at,
+        completed_at: job.completed_at,
+      };
+
+      const jobStartTime = job.started_at ? new Date(job.started_at) : new Date(0);
+      const jobEndTime = job.completed_at ? new Date(job.completed_at) : new Date(0);
+      const jobDurationMs = jobEndTime.getTime() - jobStartTime.getTime();
+
+      client.trackEvent({
+        name: `Job: ${job.name}`,
+        properties: {
+          ...jobMetadata,
+          duration_ms: jobDurationMs,
+          startTime: jobStartTime.toISOString(),
+          endTime: jobEndTime.toISOString(),
+        },
+      });
+
+
+      // Iterate over the steps in each job and track step information
+      if (isLastJob) {
```

Comment:
Shouldn't be a need for this if as I think it is always true at this point

Reply:
oh yeah and there's probably more dumb stuff there as well - i'm coding with my TS hammer